File: internal/runner/runner.go
package runner

import (
	"context"
	"errors"
	"fmt"
	"os/exec"
	"strings"
	"syscall"
	"time"
)

const (
	DefaultTimeout = 2 * time.Minute
	maxOutput      = 64 * 1024
)

type Result struct {
	ExitCode int
	Output   string
	TimedOut bool
}

// Run executes command through `sh -c` in dir, capturing combined
// stdout/stderr (tail-capped at maxOutput). The whole process group is
// killed when timeout elapses.
func Run(ctx context.Context, dir, command string, timeout time.Duration) (Result, error) {
	if timeout <= 0 {
		timeout = DefaultTimeout
	}
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "sh", "-c", command)
	cmd.Dir = dir
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	cmd.Cancel = func() error {
		return syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
	}
	cmd.WaitDelay = 5 * time.Second

	out := &tailBuffer{limit: maxOutput}
	cmd.Stdout = out
	cmd.Stderr = out

	err := cmd.Run()
	result := Result{Output: out.String(), TimedOut: ctx.Err() == context.DeadlineExceeded}
	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			result.ExitCode = exitErr.ExitCode()
			return result, nil
		}
		if result.TimedOut {
			result.ExitCode = -1
			return result, nil
		}
		return result, fmt.Errorf("running %q: %w", command, err)
	}
	return result, nil
}

// tailBuffer keeps only the last `limit` bytes written to it.
type tailBuffer struct {
	limit     int
	buf       []byte
	truncated bool
}

func (b *tailBuffer) Write(p []byte) (int, error) {
	b.buf = append(b.buf, p...)
	if len(b.buf) > b.limit {
		b.buf = b.buf[len(b.buf)-b.limit:]
		b.truncated = true
	}
	return len(p), nil
}

func (b *tailBuffer) String() string {
	s := strings.TrimSpace(string(b.buf))
	if b.truncated {
		return "[output truncated, showing the tail]\n" + s
	}
	return s
}

File: src/nina/system/runner.py
from __future__ import annotations

import asyncio
import contextlib
import os
import signal
from dataclasses import dataclass

DEFAULT_TIMEOUT = 120.0
MAX_OUTPUT = 64 * 1024


@dataclass
class Result:
    exit_code: int
    output: str
    timed_out: bool


class TailBuffer:
    def __init__(self, limit: int) -> None:
        self._limit = limit
        self._buf = bytearray()
        self._truncated = False

    def write(self, data: bytes) -> None:
        self._buf += data
        if len(self._buf) > self._limit:
            self._buf = self._buf[-self._limit :]
            self._truncated = True

    def __str__(self) -> str:
        s = self._buf.decode(errors="replace").strip()
        if self._truncated:
            return "[output truncated, showing the tail]\n" + s
        return s


async def run(dir: str, command: str, timeout: float | None = None) -> Result:
    if not timeout or timeout <= 0:
        timeout = DEFAULT_TIMEOUT

    proc = await asyncio.create_subprocess_shell(
        command,
        cwd=dir,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.STDOUT,
        preexec_fn=os.setsid,
    )

    out = TailBuffer(MAX_OUTPUT)
    timed_out = False

    async def pump() -> None:
        assert proc.stdout is not None
        while True:
            chunk = await proc.stdout.read(4096)
            if not chunk:
                return
            out.write(chunk)

    pump_task = asyncio.create_task(pump())
    try:
        await asyncio.wait_for(asyncio.shield(proc.wait()), timeout=timeout)
    except TimeoutError:
        timed_out = True
        with contextlib.suppress(ProcessLookupError):
            os.killpg(proc.pid, signal.SIGKILL)
        await proc.wait()
    await pump_task

    exit_code = -1 if timed_out else (proc.returncode or 0)
    return Result(exit_code=exit_code, output=str(out), timed_out=timed_out)
